src/server/worker.rs

```diff
@@ -2,75 +2,146 @@ use std::sync::mpsc::{Receiver, Sender};
+            Status::Running => "running",
+            Status::Failing => "failing",
+            Status::Finished => "finished",
+            Status::Unknown => "unknown",
+        };
+        write!(f, "{}", name)
+    }
+}
+
+struct MonitorTask {
+    name: String,
+    childs: Vec<Child>,
+    finished: Vec<(u32, ExitStatus)>,
+    status: Status,
 }
 
-fn monitor(m: Arc<Mutex<Vec<Child>>>, sender: Sender<Action>) {
+impl MonitorTask {
+    fn new(name: String, childs: Vec<Child>) -> Self {
+        MonitorTask {
+            name,
+            childs,
+            finished: Vec::new(),
+            status: Status::Running,
+        }
+    }
+}
+
+fn monitor(m: Arc<Mutex<MonitorTask>>, _sender: &Sender<Status>) {
     thread::spawn(move || {
         let delay: Duration = Duration::from_secs(10);
-        let mut finished: Vec<u32> = Vec::new();
+        // let mut finished: Vec<u32> = Vec::new();
         log::debug!("Start monitoring");
         loop {
-            let mut jobs = m.lock().unwrap();
+            let mut mon = m.lock().unwrap();
+            let name = mon.name.to_string();
+            let mut finished = mon.finished.clone();
 
-            for child in &mut *jobs {
+            for child in &mut mon.childs {
                 match child.try_wait() {
                     Ok(Some(status)) => {
-                        log::debug!("Exited with status {}", status);
-                        finished.push(child.id())
+                        log::debug!("[{}] child exited with status {}", name, status);
+                        finished.push((child.id(), status))
                     }
                     Ok(None) => {
-                        log::debug!("Not finished yet!");
+                        log::debug!("[{}] Not finished yet!", name);
                     }
                     Err(_) => {
-                        log::debug!("Something went wrong");
+                        log::debug!("[{}] Something went wrong", name);
                     }
                 }
             }
-            jobs.retain(|child| finished.iter().any(|&done| done == child.id()));
-            finished.clear();
-            if jobs.is_empty() {
-                log::debug!("Finished!");
-                sender.send(Action::Finished).unwrap();
+            mon.childs
+                .retain(|child| !finished.iter().any(|&(id, _st)| id == child.id()));
+            mon.finished = finished;
+            if mon.childs.is_empty() {
+                log::debug!("[{}] Finished!", name);
+                mon.status = Status::Finished;
                 break;
             }
-            drop(jobs);
-            log::debug!("Went to sleep");
+            drop(mon);
+            log::debug!("[{}] Went to sleep", name);
             thread::sleep(delay);
         }
     });
 }
 
-pub fn run(task: Task, sender: Sender<Action>, receiver: Receiver<Action>) {
+pub fn run(taskname: &str, task: Task, sender: Sender<Status>, receiver: Receiver<Action>) {
```

Comment:
I'm not sure, but it feels like the loop is infinite if we don't send an Action::Stop and we don't send one when the task is finished.